File: src/parseCli.ts
import type { CustomArgs, CustomArgValue, ParsedCli } from './types.js';

const DELIMITER = '--';
const ARG_PREFIX = '--';
const DEFAULT_PLAYWRIGHT_ARGS: readonly string[] = ['test'];

export interface SplitArgv {
  customArgv: string[];
  forwardedArgv: string[];
}

/** Splits argv at the first `--` into custom arguments and arguments forwarded to Playwright. */
export function splitArgv(argv: readonly string[]): SplitArgv {
  const delimiterIndex = argv.indexOf(DELIMITER);
  if (delimiterIndex === -1) return { customArgv: [...argv], forwardedArgv: [] };

  return {
    customArgv: argv.slice(0, delimiterIndex),
    forwardedArgv: argv.slice(delimiterIndex + 1),
  };
}

export function parseCli(argv: readonly string[]): ParsedCli {
  const { customArgv, forwardedArgv } = splitArgv(argv);

  return {
    customArgs: parseCustomArgs(customArgv),
    playwrightArgs: forwardedArgv.length > 0 ? forwardedArgv : [...DEFAULT_PLAYWRIGHT_ARGS],
  };
}

function parseCustomArgs(argv: readonly string[]): CustomArgs {
  // A Map keeps names such as "__proto__" as plain data instead of touching the object prototype.
  const result = new Map<string, CustomArgValue>();

  for (const arg of argv) {
    const { name, value } = parseCustomArg(arg);
    result.set(name, mergeValue(result.get(name), value));
  }

  return Object.fromEntries(result);
}

function parseCustomArg(arg: string): { name: string; value: string | true } {
  if (!arg.startsWith(ARG_PREFIX)) throw new Error(`Custom argument must start with "--": ${arg}`);

  const body = arg.slice(ARG_PREFIX.length);
  const equalsIndex = body.indexOf('=');
  const name = equalsIndex === -1 ? body : body.slice(0, equalsIndex);

  if (!name) throw new Error('Custom argument name cannot be empty');

  return { name, value: equalsIndex === -1 ? true : body.slice(equalsIndex + 1) };
}

function mergeValue(previous: CustomArgValue | undefined, value: string | true): CustomArgValue {
  if (previous === undefined) return value;
  if (Array.isArray(previous)) return [...previous, String(value)];
  return [String(previous), String(value)];
}

File: src/parse_cli.rs
//! Port of `parseCli.ts`.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

/// A custom argument value: `--flag` (bool), `--name=value` (string) or repeats (list).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ArgValue {
    Bool(bool),
    Str(String),
    List(Vec<String>),
}

impl ArgValue {
    /// Mirrors JavaScript `String(value)` coercion.
    pub fn into_string(self) -> String {
        match self {
            ArgValue::Bool(value) => value.to_string(),
            ArgValue::Str(value) => value,
            ArgValue::List(values) => values.join(","),
        }
    }
}

impl From<&str> for ArgValue {
    fn from(value: &str) -> Self {
        ArgValue::Str(value.to_string())
    }
}

impl From<bool> for ArgValue {
    fn from(value: bool) -> Self {
        ArgValue::Bool(value)
    }
}

impl<const N: usize> From<[&str; N]> for ArgValue {
    fn from(values: [&str; N]) -> Self {
        ArgValue::List(values.iter().map(|value| value.to_string()).collect())
    }
}

/// Insertion-ordered map, matching JavaScript object key order.
pub type CustomArgs = IndexMap<String, ArgValue>;

#[derive(Debug, PartialEq)]
pub struct ParsedCli {
    pub custom_args: CustomArgs,
    pub playwright_args: Vec<String>,
}

pub fn parse_cli(argv: &[String]) -> Result<ParsedCli, String> {
    let delimiter_index = argv.iter().position(|arg| arg == "--");
    let custom_argv = &argv[..delimiter_index.unwrap_or(argv.len())];
    let forwarded = delimiter_index.map_or(&[] as &[String], |index| &argv[index + 1..]);

    Ok(ParsedCli {
        custom_args: parse_custom_args(custom_argv)?,
        playwright_args: if forwarded.is_empty() {
            vec!["test".to_string()]
        } else {
            forwarded.to_vec()
        },
    })
}

fn parse_custom_args(argv: &[String]) -> Result<CustomArgs, String> {
    let mut result = CustomArgs::new();

    for arg in argv {
        let without_prefix = arg
            .strip_prefix("--")
            .ok_or_else(|| format!("Custom argument must start with \"--\": {arg}"))?;

        let (name, value) = match without_prefix.split_once('=') {
            None => (without_prefix, ArgValue::Bool(true)),
            Some((name, value)) => (name, ArgValue::Str(value.to_string())),
        };

        if name.is_empty() {
            return Err("Custom argument name cannot be empty".to_string());
        }

        match result.get_mut(name) {
            None => {
                result.insert(name.to_string(), value);
            }
            Some(ArgValue::List(previous)) => previous.push(value.into_string()),
            Some(previous) => {
                let first = std::mem::replace(previous, ArgValue::Bool(true)).into_string();
                *previous = ArgValue::List(vec![first, value.into_string()]);
            }
        }
    }

    Ok(result)
}
